onnx/defs/math/defs.cc

```diff
@@ -2586,73 +2586,121 @@ void einsumRankInference(ONNX_NAMESPACE::InferenceContext& ctx, std::string equa
 
   // Parse the left-hand side
   std::stringstream str(left_equation);
+  std::map<char, size_t> label_maps;
+  std::set<char> repeated_labels;
+  ONNX_NAMESPACE::TensorShapeProto dims_value, ellipsis_dims_value;
+  size_t num_labels = 0;
+  bool ellipsis_flag = true;
+
   while (!str.eof()) {
     std::getline(str, term, ',');
     auto ellipsis_index = term.find("...");
-    if (numInputs <= num_operands) {
+    if (num_inputs <= num_operands) {
       fail_shape_inference("Number of input tensors does not match the operands in the equation.");
     }
-    size_t rank = ctx.getInputType(num_operands)->tensor_type().shape().dim_size();
+    const auto& shape = ctx.getInputType(num_operands)->tensor_type().shape();
+    size_t rank = shape.dim_size();
+    size_t ellipsis_dims = 0;
+
+    size_t term_size = 0; // number of legal indices for the current term
+    size_t num_illegal_char = 0; // number of illegal char before the current 'index' in the current term
+
+    for (size_t index = 0; index < term.size(); ++index) {
+      if (is_letter(term[index])) {
+        term_size += 1;
+      }
+    }
+
+    for (size_t index = 0; index < term.size(); ++index) {
+      if (index == ellipsis_index) {
+        // find ellipsis and record the dims represented by ellipsis
+        ellipsis_dims = rank - term_size;
+        if (ellipsis_flag) {
```

Comment:
I am not sure this is correct. The op spec says that ellipsis dimensions are broadcast ... so the first ellipsis dimensions may be [1] and the second might be [32] ... then shouldn't the output ellipsis dimension be [32]?

Reply:
Op spec says that "The equation may contain ellipsis (“…”) to enable broadcasting. Ellipsis must indicate a fixed number of dimensions. Specifically, every occurrence of ellipsis in the equation must represent the same number of dimensions." The ellipsis may appear more than once in the left part of the equation (i.e., it can occur in more than one term.) We should only record the dimensions represented by the first ellipsis. 
For example, "abij,abjk->abik" can be rewritten as "...ij,...jk->...ik". "..." represents dimensions "ab".

Reply:
Every occurrence of ellipsis must represent the same number of dimensions, that is correct. But the values of those dimensions don't need to match ... at least, that's my interpretation of "broadcasting" here. So, I interpret "...ij, ...jk -> ...ik" to mean that inputs of shape [1, M, N] and [32, N, K] will produce an output of shape [32, M, K] ... isn't that right?

Reply:
Oh, sorry, I misunderstood this paragraph... I will fix this error.